File: web/packages/alloy-storage/src/backends/drive/drive-storage.ts
import { GoogleAuth, type GoogleAuthDeps } from '../../auth/google-auth.js';
import { DriveBackend } from './drive-backend.js';
import { DriveClient } from './drive-client.js';

const DEFAULT_SCOPE = 'https://www.googleapis.com/auth/drive.file';

export interface DriveStorageConfig {
  clientId: string;
  redirectUri: string;
  tokenServiceUrl: string;
  /** Folder path find-or-created from the Drive root, e.g. 'AllyWorld/AllyClock'. */
  folderPath: string;
  /** Defaults to drive.file — the app sees only files it created. */
  scope?: string;
}

/** Injection seams, forwarded to the underlying pieces (tests, custom transports). */
export interface DriveStorageDeps {
  auth?: GoogleAuthDeps;
  fetchFn?: typeof fetch;
  cache?: Storage | null;
}

export interface DriveStorage {
  auth: GoogleAuth;
  backend: DriveBackend;
}

/** One-call wiring of the Drive stack: GoogleAuth → DriveClient → DriveBackend.
 *  The client is internal plumbing; apps keep the two objects they use.
 *  Sugar, not a seal — the individual constructors remain public. */
export function createDriveStorage(
  config: DriveStorageConfig,
  deps: DriveStorageDeps = {}
): DriveStorage {
  const auth = new GoogleAuth(
    {
      clientId: config.clientId,
      scope: config.scope ?? DEFAULT_SCOPE,
      redirectUri: config.redirectUri,
      tokenServiceUrl: config.tokenServiceUrl,
    },
    deps.auth ?? {}
  );
  const client = deps.fetchFn ? new DriveClient(auth, deps.fetchFn) : new DriveClient(auth);
  const backend =
    deps.cache === undefined
      ? new DriveBackend(client, config.folderPath)
      : new DriveBackend(client, config.folderPath, deps.cache);
  return { auth, backend };
}

File: swift/Sources/AlloyStorage/Backends/Drive/DriveStorage.swift
import Foundation

/// Config for the one-call Drive stack. Twin of DriveStorageConfig in
/// drive-storage.ts — fields differ per platform exactly as GoogleAuthConfig
/// does (web: redirectUri + tokenServiceUrl; Apple: redirectScheme).
public struct DriveStorageConfig: Sendable {
  public let clientId: String
  public let redirectScheme: String
  /// Folder path find-or-created from the Drive root, e.g. "AllyWorld/AllyClock".
  public let folderPath: String
  /// Defaults to drive.file — the app sees only files it created.
  public let scope: String

  public init(
    clientId: String, redirectScheme: String, folderPath: String,
    scope: String = "https://www.googleapis.com/auth/drive.file"
  ) {
    self.clientId = clientId
    self.redirectScheme = redirectScheme
    self.folderPath = folderPath
    self.scope = scope
  }
}

/// One-call wiring of the Drive stack: GoogleAuth → DriveClient → DriveBackend.
/// The client is internal plumbing; apps keep the two objects they use.
/// Sugar, not a seal — the individual initializers remain public.
public struct DriveStorage {
  public let auth: GoogleAuth
  public let backend: DriveBackend

  /// `uiSession` mirrors `GoogleAuth.init`'s default exactly: the platform
  /// auth UI (`GoogleAuth.defaultUISession()`), not `nil` — passing `nil`
  /// explicitly here means "no auth UI," same as it does for `GoogleAuth`
  /// itself, so a caller who omits the argument still gets working sign-in.
  public init(
    config: DriveStorageConfig,
    vault: any TokenVault = KeychainTokenVault(),
    transport: any HTTPTransport = URLSessionTransport(),
    uiSession: (any AuthUISession)? = GoogleAuth.defaultUISession(),
    cache: sending UserDefaults? = .standard
  ) {
    let auth = GoogleAuth(
      config: GoogleAuthConfig(
        clientId: config.clientId, scope: config.scope, redirectScheme: config.redirectScheme),
      vault: vault, transport: transport, uiSession: uiSession)
    self.auth = auth
    self.backend = DriveBackend(
      client: DriveClient(auth: auth, transport: transport),
      folderPath: config.folderPath, cache: cache)
  }
}
